pkg/instancetype/instancetype.go

```diff
@@ -521,6 +531,123 @@ func (m *methods) findClusterInstancetype(vm *virtv1.VirtualMachine) (*instancet
+	if err != nil {
+		return "", "", err
+	}
+	return inferDefaultsFromAnnotations(pvc.Annotations, defaultNameAnnotation, defaultKindAnnotation)
+}
+
+func (m *methods) inferDefaultsFromDataVolume(vm *virtv1.VirtualMachine, dvName, defaultNameAnnotation, defaultKindAnnoation string) (string, string, error) {
+	if len(vm.Spec.DataVolumeTemplates) > 0 {
+		for _, dvt := range vm.Spec.DataVolumeTemplates {
+			if dvt.Name != dvName {
+				continue
+			}
+			return m.inferDefaultsFromDataVolumeSpec(dvt.Spec, defaultNameAnnotation, defaultKindAnnoation)
+		}
+	}
+	dv, err := m.clientset.CdiClient().CdiV1beta1().DataVolumes(vm.Namespace).Get(context.Background(), dvName, metav1.GetOptions{})
+	if err != nil {
+		return "", "", err
+	}
+	// Check the DataVolume for any annotations before checking the underlying PVC
+	defaultName, defaultKind, err := inferDefaultsFromAnnotations(dv.Annotations, defaultNameAnnotation, defaultKindAnnoation)
+	if err == nil {
+		return defaultName, defaultKind, nil
+	}
+	return m.inferDefaultsFromDataVolumeSpec(dv.Spec, defaultNameAnnotation, defaultKindAnnoation)
+}
+
+func (m *methods) inferDefaultsFromDataVolumeSpec(dataVolumeSpec v1beta1.DataVolumeSpec, defaultNameAnnotation, defaultKindAnnoation string) (string, string, error) {
+	if dataVolumeSpec.Source != nil && dataVolumeSpec.Source.PVC != nil {
+		return m.inferDefaultsFromPVC(dataVolumeSpec.Source.PVC.Name, dataVolumeSpec.Source.PVC.Namespace, defaultNameAnnotation, defaultKindAnnoation)
+	}
+	if dataVolumeSpec.SourceRef != nil {
+		return m.inferDefaultsFromDataVolumeSourceRef(dataVolumeSpec.SourceRef, defaultNameAnnotation, defaultKindAnnoation)
+	}
+	return "", "", fmt.Errorf("unable to infer defaults from DataVolumeSpec as DataVolumeSource is not supported")
+}
+
+func (m *methods) inferDefaultsFromDataSource(dataSourceName, dataSourceNamespace, defaultNameAnnotation, defaultKindAnnoation string) (string, string, error) {
+	ds, err := m.clientset.CdiClient().CdiV1beta1().DataSources(dataSourceNamespace).Get(context.Background(), dataSourceName, metav1.GetOptions{})
+	if err != nil {
+		return "", "", err
+	}
+	// Check the DataSource for any annotations before checking the underlying PVC
+	defaultName, defaultKind, err := inferDefaultsFromAnnotations(ds.Annotations, defaultNameAnnotation, defaultKindAnnoation)
+	if err == nil {
+		return defaultName, defaultKind, nil
+	}
+	if ds.Spec.Source.PVC != nil {
+		return m.inferDefaultsFromPVC(ds.Spec.Source.PVC.Name, ds.Spec.Source.PVC.Namespace, defaultNameAnnotation, defaultKindAnnoation)
+	}
+	return "", "", fmt.Errorf("unable to infer defaults from DataSource that doesn't provide DataVolumeSourcePVC")
+}
+
+func (m *methods) inferDefaultsFromDataVolumeSourceRef(sourceRef *v1beta1.DataVolumeSourceRef, defaultNameAnnotation, defaultKindAnnoation string) (string, string, error) {
+	switch sourceRef.Kind {
+	case "PersistentVolumeClaim":
```

Comment:
Are there constants for `PersistentVolumeClaim` and `DataSource` available?

Reply:
*facepalm* that makes things slightly easier, nice catch.